dns_server/src/crawler_p2p/crawler/address_data.rs

```diff
@@ -0,0 +1,326 @@
+pub enum AddressState {
+    Connecting {
+        /// The number of consecutive failed connection attempts.
+        /// New connection attempts are made after a progressive backoff time.
+        fail_count: u32,
+
+        /// Whether the address was reachable at least once.
+        /// Addresses that were once reachable are stored in the DB.
+        was_reachable: bool,
+    },
+
+    Connected {},
+
+    Disconnecting {
+        /// Same as above
+        fail_count: u32,
+
+        /// Same as above
+        was_reachable: bool,
+    },
+
+    Disconnected {
+        /// Same as above
+        fail_count: u32,
+
+        /// Same as above
+        was_reachable: bool,
+
+        /// The time when the address went into the disconnected state
+        disconnected_at: Duration,
+    },
+
+    PermanentlyUnreachable {
+        /// Same as above
+        fail_count: u32,
+
+        /// Same as above
+        was_reachable: bool,
+
+        /// At which time the address would be removed from memory
+        erase_after: Duration,
+    },
+}
+
+/// Additional state of a potential node address
+pub struct AddressData {
+    /// Connection state
+    pub state: AddressState,
+
+    /// Whether the address was added from the command line
+    pub user_added: ConstValue<bool>,
+}
+
+impl AddressState {
+    fn fail_count(&self) -> u32 {
+        match self {
+            AddressState::Connecting {
+                fail_count,
+                was_reachable: _,
+            } => *fail_count,
+            AddressState::Connected {} => 0,
+            AddressState::Disconnecting {
+                fail_count,
+                was_reachable: _,
+            } => *fail_count,
+            AddressState::Disconnected {
+                fail_count,
+                was_reachable: _,
+                disconnected_at: _,
+            } => *fail_count,
+            AddressState::PermanentlyUnreachable {
+                fail_count,
+                erase_after: _,
+                was_reachable: _,
+            } => *fail_count,
+        }
+    }
+
+    fn was_reachable(&self) -> bool {
+        match self {
+            AddressState::Connecting {
+                fail_count: _,
+                was_reachable,
+            } => *was_reachable,
+            AddressState::Connected {} => true,
+            AddressState::Disconnecting {
+                fail_count: _,
+                was_reachable,
+            } => *was_reachable,
+            AddressState::Disconnected {
+                fail_count: _,
+                was_reachable,
+                disconnected_at: _,
+            } => *was_reachable,
+            AddressState::PermanentlyUnreachable {
+                erase_after: _,
+                was_reachable,
+                fail_count: _,
+            } => *was_reachable,
+        }
+    }
+
+    /// Whether the address is currently recognized as reachable (available from DNS)
+    pub fn is_reachable(&self) -> bool {
```

Comment:
is_connected?

Reply:
We can change the logic later to mark the node as reachable even if we are not currently connected. So I think `is_persistent` is a better name. This is only used in `CrawlerManager`.